### PR TITLE
[On-Chain-Government] Community Fund Request proposal

### LINKED ISSUE #534: [On-Chain Governance] Community fund proposal
#### What would you like to be added:

As part of #344

### PR DESCRIPTION
**Pull request by bvbfan**
/kind feature

Fixes #534 #564

### CODE REVIEW

File: src/chainparams.cpp [hunk 1]
```diff
@@ -182,6 +182,15 @@ class CMainParams : public CChainParams {
         consensus.spv.anchors_address = "1FtZwEZKknoquUb6DyQHFZ6g6oomXJYEcb";
         consensus.spv.minConfirmations = 6;
 
+        consensus.props.cfp.fee = 10 * COIN;
+        consensus.props.cfp.majorityThreshold = 5000; // vote pass with over 50% majority
+        consensus.props.voc.fee = 50 * COIN;
+        consensus.props.voc.majorityThreshold = 6667; // vote pass with over 66.67% majority
+        consensus.props.brp.fee = 500 * COIN;
+        consensus.props.brp.majorityThreshold = 6667; // vote pass with over 66.67% majority
+        consensus.props.minVoting = 1000; // 10% of the masternodes must vote
```
Comment (uzyn) [inline src/chainparams.cpp:191]: We're not implementing this rule.

File: src/chainparams.cpp [hunk 2]
```diff
@@ -182,6 +182,15 @@ class CMainParams : public CChainParams {
...
+        consensus.props.votingPeriod = 70000; // tally votes every 70K blocks
```
Comment (uzyn) [inline src/chainparams.cpp:192]: Please change this to 90k, as now our block time is 30s.

File: src/masternodes/proposals.h [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+std::string CPropTypeToString(const CPropType status);
+std::string CPropVoteToString(const CPropVoteType status);
+std::string CPropStatusToString(const CPropStatusType status);
+
+struct CCreatePropMessage {
+    CPropType type;
+    CScript address;
+    CAmount nAmount;
+    uint8_t nCycles;
+    std::string title;
```
Comment (uzyn) [inline src/masternodes/proposals.h:45]: Is there a limit on number of chars? I would propose a limit here, perhaps 256 chars.

**Review by uzyn — CHANGES_REQUESTED** [review]
Looks good so far. Looking forward to the tests on votes tallying and payout.

File: src/masternodes/proposals.h (hunk 3 above)
Comment (bvbfan) [reply src/masternodes/proposals.h:45]: Limit is 128 chars

File: src/masternodes/rpc_proposals.cpp [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+    if (!data["title"].isNull()) {
+        title = data["title"].get_str();
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "<title> is required");
+    }
+
+    if (!data["cycles"].isNull()) {
+        cycles = data["cycles"].get_int();
+        if (cycles > 3 || cycles <= 0) {
```
Comment (ShengguangXiao) [inline src/masternodes/rpc_proposals.cpp:98]: Can define 3 as a const variable?

File: src/chainparams.cpp (hunk 1 above)
Comment (bvbfan) [reply src/chainparams.cpp:191]: Here is https://github.com/DeFiCh/ain/pull/541/commits/f39815f8d3bff7a41fcfba20cb8bdef231ce719a#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R2873-R2988
On every end cycle height it's calculated active masternodes to it, then it's calculated against min voting and votes yes against all votes.

File: src/masternodes/rpc_proposals.cpp [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+
+    if (!data["title"].isNull()) {
+        title = data["title"].get_str();
+    } else {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "<title> is required");
+    }
+
+    if (!data["cycles"].isNull()) {
+        cycles = data["cycles"].get_int();
+        if (cycles > 3 || cycles < 1) {
```
Comment (Bushstar) [inline src/masternodes/rpc_proposals.cpp:98]: Should be int(MAX_CYCLES) instead of 3.

File: src/masternodes/proposals.h [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+{
+public:
+    ~CPropsView() override = default;
+
+    Res CreateProp(const CPropId& propId, uint32_t height, const CCreatePropMessage& prop);
+    Optional<CPropObject> GetProp(const CPropId& propId);
+    Res UpdatePropCycle(const CPropId& propId, uint8_t cycle);
+    Res UpdatePropStatus(const CPropId& propId, uint32_t height, CPropStatusType status);
+    Res AddPropVote(const CPropId& propId, const uint256& masternodeId, CPropVoteType vote);
+    Optional<CPropVoteType> GetPropVote(const CPropId& propId, uint8_t cycle, const uint256& masternodeId);
```
Comment (Bushstar) [inline src/masternodes/proposals.h:139]: `GetPropVote` is currently unused.

Comment (bvbfan) [reply src/masternodes/proposals.h:139]: Yes, it's still unused. Probably in the future.

File: src/masternodes/mn_checks.cpp [hunk 7]
```diff
@@ -454,6 +475,14 @@ class CCustomTxVisitor : public boost::static_visitor<Res>
         return Res::Ok();
     }
 
+    Res CheckProposalTx(CPropType type) const {
+        if (tx.vout.size() < 1
```
Comment (Bushstar) [inline src/masternodes/mn_checks.cpp:479]: No need to check for 0 vouts as this throws a ""bad-txns-vout-empty" error.

File: src/masternodes/mn_checks.cpp [hunk 8]
```diff
@@ -1904,7 +1904,10 @@ Res ApplyCustomTx(CCustomCSView& mnview, const CCoinsViewCache& coins, const CTr
         res = CustomTxVisit(view, coins, tx, height, consensus, txMessage, time);
 
         // Track burn fee
-        if (txType == CustomTxType::CreateToken || txType == CustomTxType::CreateMasternode) {
+        if (txType == CustomTxType::CreateToken
+        || txType == CustomTxType::CreateMasternode
+        || (height >= uint32_t(consensus.FortCanningHeight)
+        && txType == CustomTxType::CreateCfp)) {
```
Comment (Bushstar) [inline src/masternodes/mn_checks.cpp:1910]: Make sure that this shows as expected in getburninfo.

**Review by Bushstar — APPROVED** [review]

**Pull request merged**